feat: add ExperimentContext and e2e examples

QuentinBrosse opened the pull request:
- RecordExperimentExample -> repro of langsmith-rest-experiments repo
- E2eEvalExample -> e2e (real OpenAI agent -> dataset -> examples -> experiments -> LLM-as-a-juge)
- Some improvements in wappers

# Todo:
- [x] Rework `ExperimentContext`?
- [x] Full review


Close REL-562 REL-563

QuentinBrosse committed:
doc: RecordExperiment

QuentinBrosse committed:
doc: add eval examples

QuentinBrosse committed:
fix

QuentinBrosse committed:
improve

QuentinBrosse commented on langsmith-java-example/src/main/kotlin/com/langchain/smith/example/E2eEvalExample.kt:
```diff
@@ -0,0 +1,290 @@
+        try {
+            // Use try-with-resources to automatically manage context lifecycle
+            // This is the recommended pattern - context is automatically cleared when done
+            ExperimentContext.withExperiment(example.id(), dataset.id(), sessionId).use {
```
Check if we need the dataset_id here – we already have the example_id.

QuentinBrosse committed:
improve

QuentinBrosse committed:
improve

lunevalex commented on langsmith-java-core/src/main/java/com/langchain/smith/wrappers/openai/ExperimentContext.java:
```diff
@@ -0,0 +1,244 @@
+     * @return a map of metadata key-value pairs
+     */
+    public static Map<String, String> getMetadata() {
+        return new HashMap<>(CONTEXT.get().metadata);
```
Collections.unmodifiableMap() is a better pattern

lunevalex commented on langsmith-java-core/src/main/java/com/langchain/smith/wrappers/openai/ExperimentContext.java:
```diff
@@ -0,0 +1,244 @@
+ */
+public final class ExperimentContext {
+
+    private static final ThreadLocal<Context> CONTEXT = ThreadLocal.withInitial(Context::new);
```
I would make this class a singleton, rather than static.

QuentinBrosse replied:
I went with static because it follows the same pattern as OpenTelemetry's Context:

  ```java
  // OpenTelemetry does this (static):
  Context.current().with(key, value).makeCurrent()

  // Not this (singleton):
  Context.getInstance().current().with(key, value).makeCurrent()
  ```
  
Do you see a specific benefit to the singleton pattern here that I'm missing?

QuentinBrosse replied:
but still refactored to match OpenTelemetry Context pattern

lunevalex commented on langsmith-java-core/src/main/java/com/langchain/smith/wrappers/openai/OpenTelemetryConfig.java:
```diff
@@ -65,6 +71,234 @@ public enum SpanProcessorType {
+         * @return the configured OpenTelemetry instance
+         * @throws IllegalArgumentException if apiKey is not set
+         */
+        public OpenTelemetry build() {
```
probably want to do some validation here and throw up if the parameters are missing, instead of getting an NPE down the line.

QuentinBrosse replied:
yeah I had a ticket to refactor that but done it here

lunevalex commented on langsmith-java-example/src/main/kotlin/com/langchain/smith/example/E2eEvalExample.kt:
```diff
@@ -0,0 +1,283 @@
+        println("   ✓ Created new dataset (ID: ${created.id()})")
+        created
+    }
+    println("   → View dataset: ${buildDatasetUrl(dataset)}")
```
I'd use the logger "java.util.logging", I know its example code but still

QuentinBrosse replied:
I would keep print in examples as I feel it's better for educational code.

However, we now use SLF4J in all production code, allowing customers to choose which backend they want and keep deps under control

lunevalex commented on langsmith-java-core/src/main/java/com/langchain/smith/wrappers/openai/ExperimentContext.java:
```diff
@@ -0,0 +1,244 @@
+    /**
+     * Internal context holder for thread-local data.
+     */
+    private static class Context {
```
its an internal class, but still kind of a java smell to have direct field access.

lunevalex commented on langsmith-java-core/src/main/java/com/langchain/smith/wrappers/openai/OpenTelemetryConfig.java:
```diff
@@ -65,6 +71,234 @@ public enum SpanProcessorType {
+     *     .build();
+     * }</pre>
+     */
+    public static class Builder {
```
we could probably just use something like Lombok to avoid writing it, but I guess if LLM code gen might not matter.

lunevalex commented on langsmith-java-core/src/main/java/com/langchain/smith/wrappers/openai/ExperimentContext.java:
```diff
@@ -0,0 +1,244 @@
+     * @return the reference example ID, or null if not set
+     */
+    public static String getReferenceExampleId() {
+        return CONTEXT.get().referenceExampleId;
```
Optional<> would be a good pattern here is this is nullable, makes it a lot easier to work with.

QuentinBrosse committed:
refactor: ExperimentContext to use OpenTelemetry Context pattern

QuentinBrosse committed:
refactor: migrate to SLF4J logging and Builder-only API

Refactor OpenTelemetry configuration and logging to improve usability and follow industry standards.

API Changes:
  - Remove `configureForLangSmith()` static methods in favor of Builder pattern
  - Remove `configureFromEnv()` - use `builder().build()` instead
  - Add required field validation in `Builder.build()` with clear error messages
  - Builder constructor auto-reads environment variables (LANGSMITH_API_KEY, etc.)
  - Builder methods override environment variable values

QuentinBrosse committed:
add optional + lint

QuentinBrosse committed:
some fixes

QuentinBrosse committed:
lint
